flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/matrix/SparseVector.java

```diff
@@ -0,0 +1,791 @@
+	}
+
+	public DenseVector plus(DenseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("The size of the two vectors are different.");
+		}
+
+		DenseVector r = other.clone();
+		for (int i = 0; i < this.indices.length; i++) {
+			r.add(this.indices[i], this.values[i]);
+		}
+		return r;
+	}
+
+	@Override
+	public Vector plus(Vector vec) {
+		if (vec instanceof DenseVector) {
+			return plus((DenseVector) vec);
+		} else {
+			return plus((SparseVector) vec);
+		}
+	}
+
+	public SparseVector minus(SparseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("The size of the two vectors are different.");
+		}
+
+		int totNnz = this.values.length + other.values.length;
+		int p0 = 0;
+		int p1 = 0;
+		while (p0 < this.values.length && p1 < other.values.length) {
+			if (this.indices[p0] == other.indices[p1]) {
+				totNnz--;
+				p0++;
+				p1++;
+			} else if (this.indices[p0] < other.indices[p1]) {
+				p0++;
+			} else {
+				p1++;
+			}
+		}
+
+		SparseVector r = new SparseVector(this.size());
+		r.indices = new int[totNnz];
+		r.values = new double[totNnz];
+		p0 = p1 = 0;
+		int pos = 0;
+		while (pos < totNnz) {
+			if (p0 < this.values.length && p1 < other.values.length) {
+				if (this.indices[p0] == other.indices[p1]) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0] - other.values[p1];
+					p0++;
+					p1++;
+				} else if (this.indices[p0] < other.indices[p1]) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0];
+					p0++;
+				} else {
+					r.indices[pos] = other.indices[p1];
+					r.values[pos] = -other.values[p1];
+					p1++;
+				}
+				pos++;
+			} else {
+				if (p0 < this.values.length) {
+					r.indices[pos] = this.indices[p0];
+					r.values[pos] = this.values[p0];
+					p0++;
+					pos++;
+					continue;
+				}
+				if (p1 < other.values.length) {
+					r.indices[pos] = other.indices[p1];
+					r.values[pos] = -other.values[p1];
+					p1++;
+					pos++;
+					continue;
+				}
+			}
+		}
+
+		return r;
+	}
+
+	public DenseVector minus(DenseVector other) {
+		if (this.size() != other.size()) {
+			throw new RuntimeException("The size of the two vectors are different.");
+		}
+
+		DenseVector r = other.scale(-1.0);
+		for (int i = 0; i < this.indices.length; i++) {
+			r.add(this.indices[i], this.values[i]);
+		}
+		return r;
+	}
+
+	@Override
+	public Vector minus(Vector vec) {
+		if (vec instanceof DenseVector) {
+			return minus((DenseVector) vec);
+		} else {
+			return minus((SparseVector) vec);
+		}
+	}
+
+	@Override
+	public SparseVector scale(double d) {
+		SparseVector r = new SparseVector(this.n, this.indices, this.values);
+		for (int i = 0; i < this.values.length; i++) {
+			r.values[i] *= d;
+		}
+		return r;
+	}
+
+	@Override
+	public void scaleEqual(double d) {
+		for (int i = 0; i < this.values.length; i++) {
+			this.values[i] *= d;
+		}
+	}
+
+	public void removeZeroValues() {
```

Comment:
Can we add java doc for these non-overriding public APIs?

Reply:
Thank you. We have added the java doc.